src/aotcompile.cpp

```diff
@@ -600,9 +600,12 @@ void addTargetPasses(legacy::PassManagerBase *PM, TargetMachine *TM)
 void addMachinePasses(legacy::PassManagerBase *PM, TargetMachine *TM, int optlevel)
 {
     // TODO: don't do this on CPUs that natively support Float16
-    PM->add(createDemoteFloat16Pass());
-    if (optlevel > 1)
-        PM->add(createGVNPass());
+    auto feat_string = TM->getTargetFeatureString();
+    if(feat_string.find("+fp16fml") == llvm::StringRef::npos||feat_string.find("+fp16fml") == llvm::StringRef::npos){
```

Comment:
So instead of doing this check here, I think we need to do it on a per-function basis, within the pass. 

The way this works is that the multi-versioning pass will clone the function if it contains Float16 ops and then we have two copies of the function, one that should have a target feature set and one without it. On the one that is lacking the target-feature we still want to run the DemoteFloat16 pass. 

For the JIT context there is likely a similar issue that @vtjnash, @JeffBezanson and I discussed yesterday. 
See https://github.com/JuliaLang/julia/pull/43085#discussion_r788197603 for some more context

Reply:
I moved the check to inside the pass, but I don't think the way the check is done is the best

Reply:
The multiversioning check part seems correct (per-function). The issue is `jl_ExecutionEngine->getTargetFeatureString` right now can be looking at the wrong target for many people (e.g. GPUs, static-compilation, sysimg building), which is what https://reviews.llvm.org/D107082 was hoping to solve

Reply:
Yeah, I just did it similarly to how we do FMA. Could we get that the feature flags from TTI? Or does that need your change.
Also it seems the frontend already does what demote float16 would do, but I guess it would then miss some possible optimizations

Reply:
Right now, only builtin LLVM passes are allowed to access that data, but it is forbidden to external passes. That PR would make is available to all passes, but it is possible that LLVM devs will like conceptually dislike the idea of fixing this bug.

Reply:
I guess the "correct" thing would be to add a `hasFloat16` function or something like it. But that sounds like a very roundabout way to solve something that seems quite obvious.

Reply:
Do you have any other idea of how this could be done? Also isn't it an issue if the TargetMachine disagrees with TargetFeatureString? What does machine code generation use as the truth?